ARC rewrite

Oaphi opened the pull request:
This PR completely rewrites ARC from the ground up in TypeScript

Oaphi committed:
changed to 4 spaces as indent

Oaphi committed:
removed unused project parts

Oaphi committed:
initial package config

Oaphi committed:
added testing setup (mocha + chai)

Oaphi committed:
added dependabot & GH CI workflow

Oaphi committed:
updated tscconfig

Oaphi committed:
removed old source, added relevant TS configs

Oaphi committed:
package-wide updates (dependencies)

Oaphi committed:
added variable template replacer

Oaphi committed:
first TS version without jQuery (unstable)

Oaphi committed:
switched from option template to DOM manipulation

dependabot[bot] committed:
Chores: Bump mocha from 9.0.0 to 9.0.1 (#2)

Bumps [mocha](https://github.com/mochajs/mocha) from 9.0.0 to 9.0.1.
- [Release notes](https://github.com/mochajs/mocha/releases)
- [Changelog](https://github.com/mochajs/mocha/blob/master/CHANGELOG.md)
- [Commits](https://github.com/mochajs/mocha/compare/v9.0.0...v9.0.1)

---
updated-dependencies:
- dependency-name: mocha
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Chores: Bump typescript from 4.3.2 to 4.3.4 (#3)

Bumps [typescript](https://github.com/Microsoft/TypeScript) from 4.3.2 to 4.3.4.
- [Release notes](https://github.com/Microsoft/TypeScript/releases)
- [Commits](https://github.com/Microsoft/TypeScript/compare/v4.3.2...v4.3.4)

---
updated-dependencies:
- dependency-name: typescript
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Chores: Bump @types/chai from 4.2.18 to 4.2.19 (#5)

Bumps [@types/chai](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/HEAD/types/chai) from 4.2.18 to 4.2.19.
- [Release notes](https://github.com/DefinitelyTyped/DefinitelyTyped/releases)
- [Commits](https://github.com/DefinitelyTyped/DefinitelyTyped/commits/HEAD/types/chai)

---
updated-dependencies:
- dependency-name: "@types/chai"
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Chores: Bump eslint from 7.28.0 to 7.29.0 (#4)

Bumps [eslint](https://github.com/eslint/eslint) from 7.28.0 to 7.29.0.
- [Release notes](https://github.com/eslint/eslint/releases)
- [Changelog](https://github.com/eslint/eslint/blob/master/CHANGELOG.md)
- [Commits](https://github.com/eslint/eslint/compare/v7.28.0...v7.29.0)

---
updated-dependencies:
- dependency-name: eslint
  dependency-type: direct:development
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

Oaphi committed:
Merge branch 'cleanup' of github.com:userscripters/SE-AutoReviewComments into cleanup

Oaphi committed:
second TS version without jQuery (unstable)

Oaphi committed:
moved prefix to precompiled values

Oaphi committed:
third TS version without jQuery (unstable)

Oaphi committed:
fourth TS version without jQuery (unstable)

Oaphi committed:
fifth TS version without jQuery (unstable)

Oaphi committed:
sixth TS version without jQuery (unstable)

Oaphi committed:
seventh TS version without jQuery (unstable)

Oaphi committed:
npm scripts updates

Oaphi committed:
updated positioning; grouped consts; optimized comment resetting

Oaphi committed:
replaced global id var with explicit getter call

Oaphi committed:
optimized comment type matching

Oaphi committed:
removed unnecessary IIFE since we are already sandboxed

Oaphi committed:
moved API key to utils for maintainability

Oaphi committed:
updated MD/HTML/Comment utils

Oaphi committed:
minor updates to locator funcs

Oaphi committed:
minor fixes to injectors

Oaphi committed:
updated target matching

Oaphi committed:
fixed comment type matcher

Oaphi committed:
fixed user id scraper util

Oaphi committed:
fixed quick insert buttons

Oaphi committed:
made a couple selectors configurable, minor fixes

Oaphi committed:
optimized comment loading

Oaphi committed:
removed unnecessary resetting of welcome

Oaphi committed:
reworked JSONP creation; added view switcher

Oaphi committed:
fixed instert button being enabled on comment edit cancelling

Oaphi committed:
cancelling remote/welcome edits now returns to the filter view

Oaphi committed:
improved fadeTo helper to prevent visibility togging while fading

Oaphi committed:
fixed not being able to insert comments on first open

Oaphi committed:
rebuilt dist

Oaphi committed:
Merge branch 'master' into cleanup

double-beep reviewed (changes requested):
The changes look great overall!

- I haven't really understood when/why you prefer `const name = () => {}` over `function name() {}`.
- Should we use object destructuring for just one property?

Oaphi commented on src/arc.ts:
```diff
@@ -0,0 +1,2488 @@
+     * - called at the end of the main script if function exists
+     * @param {HTMLElement} popup
+     */
+    const checkForNewVersion = (popup: HTMLElement) => {
```
yeah, we can ditch it, I just rewrote it on auto to avoid thinking too much about whether it's needed or not

double-beep replied:
I've removed it. Chrome seemed to block the requesta anyway.

Oaphi commented on src/arc.ts:
```diff
@@ -0,0 +1,2488 @@
+
+        static save(key: string, val: string | number | boolean): void {
+            const { prefix, storage } = this;
+            return void (storage[prefix + key] = JSON.stringify(val));
```
yup, that's probably an artifact from an earlier take, no `return` should suffice

Oaphi committed:
removed useless return from the Storage 'save' method

Oaphi commented on src/arc.ts:
```diff
@@ -0,0 +1,2488 @@
+     */
+    const fadeOut = (el: HTMLElement, speed = 200) => fadeTo(el, 0, speed);
+
+    class Store {
```
Agreed - I made the `Store`  easily accept other types of storage, so should be easy to support. Should I do it or will you?

Oaphi replied:
I went ahead and implemented the ability to switch to the userscript manager storage when the corresponding functions are present. The update makes it so there is a venue for making it a user preference further down the line. Also want to add support for Greasemonkey's storage functions, but the asynchronous API is killing the mood a little :) Resolved in 2c5fe61

Oaphi commented on src/arc.ts:
```diff
@@ -0,0 +1,2488 @@
+
+        log(msg: unknown, ...params: unknown[]) {
+            const { on } = this;
+            on && console.debug(msg, ...params);
```
yup, let's do that

Oaphi committed:
optimize 'remove' method call on 'classList'

the `remove` method signature allows rest parameters

Co-authored-by: double beep <38133098+double-beep@users.noreply.github.com>

Oaphi committed:
added the ARC prefix to debugger

Oaphi committed:
Merge branch 'cleanup' of github.com:userscripters/SE-AutoReviewComments into cleanup

Oaphi commented on src/arc.ts:
```diff
@@ -0,0 +1,2488 @@
+            fadeOut(seeBtn.closest(".main")!);
+        });
+
+        seeBtn.addEventListener("mouseleave", () => {
```
yeah, I thought I solved it with the locking mechanism, but I guess the implementation's a bit buggy, let's see what I can do

Oaphi replied:
done (alongside switching to CSS transitions in 7c5cab4

Oaphi commented on src/arc.ts:
```diff
@@ -0,0 +1,2488 @@
+        });
+
+        actions.append(cancel);
+        el.append(area, actions);
```
Agreed, I simply left the preview to work the same way as it did in the original ARC because of the other changes. A live preview with a `change` listener will be great - I'll see to it

Oaphi commented on src/arc.ts:
```diff
@@ -0,0 +1,2488 @@
+
+                    debugLogger.log({ op });
+
+                    insertComment(input, descr.innerHTML, op);
```
Yeah, something we need to fix. I am leaning towards simply disabling any possibility for insertion during edits

Oaphi replied:
Can you still reproduce? I disable the insertion during comment edits, so this should be fixed. Btw, do we want to allow inserting during editing? Extra complexity, but hey, I love me some.

double-beep replied:
No, this doesn't appear to be fixed for me.

Oaphi replied:
done, resolved in 31e0e7c

Oaphi commented on src/arc.ts:
```diff
@@ -0,0 +1,2488 @@
+            Target: [Target.CommentAnswer],
+            Name: "Answers just to say Thanks!",
+            Description:
+                'Please don\'t add "thanks" as answers. Invest some time in the site and you will gain sufficient <a href="//$SITEURL$/privileges">privileges</a> to upvote answers you like, which is the $SITENAME$ way of saying thank you.',
```
Agreed, didn't pay much attention to those replacements during the rewrite.

Oaphi commented:
> The changes look great overall!
> 
> * I haven't really understood when/why you prefer `const name = () => {}` over `function name() {}`.
> * Should we use object destructuring for just one property?

Thanks! As for the questions:

1. I *always* prefer arrow functions with concise/normal bodies over `function`s. If you see the latter, it means that I either forgot to change that during the rewrite or was too horrified by the Cthulhu logic there to touch it :) Consider anything `function` as an implicit TODO item.
2. Well, I use it simply for consistency because I use restructuring everywhere. As far as I recall, the overhead is insignificant, but if you want to switch to simple property access for single properties - I am good with that.

Oaphi committed:
Comment-centric updates:
- made comment links relative;
- fixed editorial mistakes in default comments;
- added inline HTML helpers;

Oaphi commented on src/arc.ts:
```diff
@@ -0,0 +1,2488 @@
+            return val ? JSON.parse(val) : def;
+        }
+
+        static save(key: string, val: string | number | boolean): void {
```
Because I thought of `load<T>` very late in the rewrite :) Thus never thought about `save` - agreed, should be done shortly

Oaphi committed:
made 'save' method generic; switched to 'setItem' instead of directly setting a property on the storage

Oaphi commented on src/arc.ts:
```diff
@@ -0,0 +1,2488 @@
+            `.${arc}.popup .main .action-list li label .quick-insert:hover{
+                    background-color:#222;
+                    color:#fff
+                }`,
+            `.${arc}.popup .actions,.auto-review-comments.popup .main .actions{
+                    margin:6px
+                }`,
+            `.${arc}.popup .main .popup-submit{
+                    float:none;
+                    margin:0 0 5px 0;
+                }`,
+            `.${arc}.announcement strong:first-child {
+                    display: block;
+                }`,
+            `.${arc}.announcement{
+                    padding:7px;
+                    margin-bottom:10px;
+                    background:orange;
+                    font-size:15px;
+                }`,
+            `.${arc}.announcement .notify-close{
+                    display:block;
+                    float:right;
+                    margin:0 4px;
+                    padding:0 4px;
+                    border:2px solid black;
+                    cursor:pointer;
+                    line-height:17px;
+                }`,
+            `.${arc}.announcement .notify-close a{
+                    color:black;
+                    text-decoration:none;
+                    font-weight:bold;
+                    font-size:16px;
+                }`,
+            `.${arc}.popup .main .searchfilter{
+                    width:100%;
+                    box-sizing:border-box;
+                    display:block
+                }`,
```
Hmmm, I thought we were going to switch to Stacks either way? So soon the whole stylesheet won't be needed.

double-beep replied:
Ok, then

Oaphi commented on src/arc.ts:
```diff
@@ -0,0 +1,2488 @@
+        showPopup(popup);
+
+        //TODO: if popup is created only once, listeners should be setup only once
+        [updateComments].forEach((initiator) => initiator(popup, postType));
```
No reason - I was gradually updating the script, so, at some point, there were multiple "initiators" (legacy of the original ARC). Forgot to get rid of it.

Oaphi committed:
remove useless forEach

Co-authored-by: double beep <38133098+double-beep@users.noreply.github.com>

Oaphi committed:
switched Target to enum

Oaphi committed:
Merge branch 'cleanup' of github.com:userscripters/SE-AutoReviewComments into cleanup

Oaphi commented on src/arc.ts:
```diff
@@ -0,0 +1,2488 @@
+        CommentAnswer: "A",
+        EditSummaryAnswer: "EA",
+        EditSummaryQuestion: "EQ",
+    };
```
Agreed, done

Oaphi commented on src/arc.ts:
```diff
@@ -0,0 +1,2488 @@
+     * @param {string} html
+     * @returns {string}
+     */
+    const HTMLtoMarkdown = (html: string) =>
```
mmm, not so sure about it. First, we have a problem with users having those comments stored, modified, etc. To use our version, they will either have to clean up should we switch or we have to add the logic that updates those users automatically. So despite me agreeing with you that would be better to store the markdown, I think we should leave this be. Re: SE - not sure this is a good idea either. I don't like relying on SE too much. Styling - sure, logic? Not so much.

double-beep replied:
I've resolved this for now. I plan to make a change so that the script only uses one property in `localStorage` similar to what [CertainPerformance did](https://github.com/CertainPerformance/SE-AutoReviewComments). Since it needs to be backward-compatible, I shall consider saving the markdown then.

Oaphi replied:
@double-beep great, thank you - I will leave the single-property change to you then. I guess I don't have a problem with the change (apart from SE functions) if we add logic for seamlessly "upgrading" the users of the old version - not sure how viable this is, though.

Oaphi commented on src/arc.ts:
```diff
@@ -0,0 +1,2488 @@
+     *
+     * @returns {Promise<HTMLElement>}
+     */
+    const fadeTo = async (
```
Yup will likely be simpler that way - no particular reason for doing this in JS other than I wanted to remind myself of how to write such a function. I will make change as soon as I have some time to spare.

Oaphi replied:
resolved in 7c5cab4

Oaphi commented on src/arc.ts:
```diff
@@ -0,0 +1,2488 @@
+     * @param {string} url resource URL
+     * @returns {Promise<object>} response
+     */
+    const getJSONP = <T>(url: string): Promise<T> =>
+        new Promise((resolve, reject) => {
+            const cbkName = `jsonp-${Date.now()}`;
+
+            const uri = new URL(url);
+            uri.searchParams.append("callback", cbkName);
+
+            const script = document.createElement("script");
+            script.src = uri.toString();
+            script.async = true;
+
+            window[cbkName] = (json: T) => {
+                script.remove();
+                delete window[cbkName];
+                resolve(json);
+            };
+
+            script.addEventListener("error", ({ error }) => {
+                script.remove();
+                delete window[cbkName];
+                reject(error);
+            });
+
+            document.body.append(script);
+        });
```
Well, it's a standard implementation of JSONP as to my knowledge - what if the remote is different? It's strange to see that SOCVR returns a response already wrapped in a callback called `callback` (effectively killing the temporary callback part). Frankly speaking, I really want to drop the whole thing and just ask the SOCVR guys to send CORS headers and JSON instead (and help out if needed).

double-beep replied:
Yes, that would be great :)

double-beep replied:
@Oaphi see [Makyen's response](https://chat.stackexchange.com/transcript/message/58544633)

Oaphi committed:
Simplified getLoggedInUserId

Oaphi commented on src/arc.ts:
```diff
@@ -0,0 +1,2488 @@
+     * @returns {string}
+     */
+    const getLoggedInUserId = () => {
+        const { options: { user: { userId } = {} } = {} } = StackExchange;
```
Yup, agreed, done

Oaphi committed:
rebuilt dist

double-beep committed:
Get rid of the auto-updater

dependabot[bot] committed:
Chores: Bump typescript from 4.3.4 to 4.3.5 (#10)

Bumps [typescript](https://github.com/Microsoft/TypeScript) from 4.3.4 to 4.3.5.
- [Release notes](https://github.com/Microsoft/TypeScript/releases)
- [Commits](https://github.com/Microsoft/TypeScript/compare/v4.3.4...v4.3.5)

---
updated-dependencies:
- dependency-name: typescript
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Chores: Bump mocha from 9.0.1 to 9.0.2 (#8)

Bumps [mocha](https://github.com/mochajs/mocha) from 9.0.1 to 9.0.2.
- [Release notes](https://github.com/mochajs/mocha/releases)
- [Changelog](https://github.com/mochajs/mocha/blob/master/CHANGELOG.md)
- [Commits](https://github.com/mochajs/mocha/compare/v9.0.1...v9.0.2)

---
updated-dependencies:
- dependency-name: mocha
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Chores: Bump @types/node from 15.12.2 to 16.0.0 (#9)

Bumps [@types/node](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/HEAD/types/node) from 15.12.2 to 16.0.0.
- [Release notes](https://github.com/DefinitelyTyped/DefinitelyTyped/releases)
- [Commits](https://github.com/DefinitelyTyped/DefinitelyTyped/commits/HEAD/types/node)

---
updated-dependencies:
- dependency-name: "@types/node"
  dependency-type: direct:development
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

Oaphi committed:
added live comment preview; fixed severe bug preventing comments from being saved

Oaphi committed:
Merge branch 'cleanup' of github.com:userscripters/SE-AutoReviewComments into cleanup

Oaphi committed:
rebuilt dist; added comment removal to TS config

Oaphi committed:
made comment list in the import/export view autoupdate

Oaphi committed:
comment preview fix; default comment typo fix

Oaphi committed:
improved MD/HTML replacers

Oaphi committed:
ensured edit mode can only be entered once

Oaphi committed:
switched fadeOut to CSS transitions

dependabot[bot] committed:
Chores: Bump eslint from 7.29.0 to 7.30.0 (#7)

Bumps [eslint](https://github.com/eslint/eslint) from 7.29.0 to 7.30.0.
- [Release notes](https://github.com/eslint/eslint/releases)
- [Changelog](https://github.com/eslint/eslint/blob/master/CHANGELOG.md)
- [Commits](https://github.com/eslint/eslint/compare/v7.29.0...v7.30.0)

---
updated-dependencies:
- dependency-name: eslint
  dependency-type: direct:development
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

Oaphi committed:
added support for userscript manager storage

Oaphi committed:
Merge branch 'cleanup' of github.com:userscripters/SE-AutoReviewComments into cleanup

double-beep committed:
FIX: missing assignment in the constructor

double-beep committed:
Replace var with const use .exec() and template strings

double-beep committed:
Revert "FIX: missing assignment in the constructor"

This reverts commit 696ab1db96af543377c057d2a614869009117d43.

Oaphi committed:
made getElementById generic

Oaphi committed:
added text input helper function

Oaphi committed:
type for the text input helper options; new helper 'getJSON' to match 'getJSONP' signature

Oaphi committed:
removed some of the unneeded debug logging

Oaphi committed:
remote URL value is now set on load

Oaphi committed:
added checkbox helper

Oaphi committed:
optimized remote view creator helper

Oaphi committed:
fixed height and scroll on overflow for comments

Oaphi committed:
Remote-centric updates:
- optimized auto-loading from remote
- fixed JSONP loading to avoid SOCVR dataset failure
- added ability to load JSON via Fetch API

Oaphi committed:
passed through default comments to reset; updated default comments structure

Oaphi committed:
made TS config more strict by checking for unused

Oaphi committed:
FIX: community wikis do not have user links

double-beep commented on src/arc.ts:
```diff
@@ -0,0 +1,2538 @@
+     * @returns {Promise<void>}
+     */
+    const loadFromRemote = async (url: string) => {
+        const isJSONP = /jsonp-data/.test(url);
```
This will only be true for requests to `comments.socvr.org`. But there are more auto comments, e.g. from [Charcoal](https://charcoal-se.org/userscripts/charcoal-autocomments/comments.js). Maybe we could initially use `fetch` and then on error fallback to JSONP.

Oaphi replied:
@double-beep yeah, I know, just wanted to test if the whole setup even works. Initially, I wanted to expand the regular expression with a *fixed* list of allowed JSONP hosts (I highly doubt there are many in use), but now I think the best solution would be to explicitly pass a parameter `isJSONP` and provide 2 separate inputs: one for JSONP sources (as is now) and another for JSON sources. In case both are specified - merge both.

Oaphi replied:
As discussed, the check is now simply dropped, as users will be able to decide for themselves which one they want to load

double-beep commented on src/arc.ts:
```diff
@@ -0,0 +1,2538 @@
+     *                            It will receive the action function as the second argument, so it know what to invoke when the "auto" link is clicked.
+     * @param {Actor} actor A function that will be called when the injected "auto" link is clicked.
+     */
+    function addTriggerButton<T extends HTMLElement>(
```
This function currently adds the "auto" link only in a post's comment section. It also needs to add that to the:

- close dialog: community-specific reason &rarr; Other
- rejection popup under "this edit causes harm"
- edit summaries of questions and answers (there are different comments for Q edit summaries and A edit summaries!)

Oaphi replied:
Yup, although, I am unsure whether I want to do that in this PR or not - aren't there too many changes already? Let's finalize this one, merge it into master, bump the version to 2.0.0, and then start building on top of it.

double-beep replied:
I'd rather we released 2.0.0 after the redesign.

Oaphi replied:
Agreed, maybe not 2.0.0, just as another PR - this one is a bit unwieldy as it is already :) I'd like to start looking at smaller diffs as soon as possible. Let's bump to 1.5.0 after merge.

Oaphi commented on src/arc.ts:
```diff
@@ -0,0 +1,2538 @@
+                Object.entries(actionMap).find(([key]) => el.matches(key)) ||
+                [];
+
+            action && action(popup, wrap);
```
no reason :) I still can't get used to it, will switch shortly

Oaphi committed:
added header autogeneration

Oaphi committed:
switched to optional chaining for func calls

Oaphi committed:
added helpers for Stacks URL inputs

Oaphi committed:
added 2 minor schema wrapping/unwrapping helpers

Oaphi committed:
passed through remote URL id to value updater

Oaphi committed:
minor optimizations re: remote sources

Oaphi committed:
unified remote input value handling; wrapped inputs in a flex container

Oaphi committed:
Remote-centric updates:
- removed old activity indicator
- added proper Stacks 'get now' buttons
- JSON & JSONP sources are saved simultaneously

Oaphi commented:
In case anyone wondering if I've seen a glaring issue with action buttons the most recent commits introduce - I know :) Over the next couple of days, I will be gradually shifting ARC to accept both JSON and JSONP simultaneously. Since that required an extra text input, I decided to start moving to Stacks earlier. Action buttons will be fixed with the next bunch of commits.

Oaphi committed:
added Stacks checkbox helper; added auto-get checkbox for the JSON source

dependabot[bot] committed:
Chores: Bump @types/greasemonkey from 4.0.1 to 4.0.2 (#11)

Bumps [@types/greasemonkey](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/HEAD/types/greasemonkey) from 4.0.1 to 4.0.2.
- [Release notes](https://github.com/DefinitelyTyped/DefinitelyTyped/releases)
- [Commits](https://github.com/DefinitelyTyped/DefinitelyTyped/commits/HEAD/types/greasemonkey)

---
updated-dependencies:
- dependency-name: "@types/greasemonkey"
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Chores: Bump ts-node from 10.0.0 to 10.1.0 (#12)

Bumps [ts-node](https://github.com/TypeStrong/ts-node) from 10.0.0 to 10.1.0.
- [Release notes](https://github.com/TypeStrong/ts-node/releases)
- [Commits](https://github.com/TypeStrong/ts-node/compare/v10.0.0...v10.1.0)

---
updated-dependencies:
- dependency-name: ts-node
  dependency-type: direct:development
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Chores: Bump @types/chai from 4.2.19 to 4.2.21 (#14)

Bumps [@types/chai](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/HEAD/types/chai) from 4.2.19 to 4.2.21.
- [Release notes](https://github.com/DefinitelyTyped/DefinitelyTyped/releases)
- [Commits](https://github.com/DefinitelyTyped/DefinitelyTyped/commits/HEAD/types/chai)

---
updated-dependencies:
- dependency-name: "@types/chai"
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

Oaphi committed:
moved view switching buttons to a button group

Oaphi committed:
Merge branch 'cleanup' of github.com:userscripters/SE-AutoReviewComments into cleanup

dependabot[bot] committed:
Chores: Bump @types/mocha from 8.2.2 to 8.2.3 (#15)

Bumps [@types/mocha](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/HEAD/types/mocha) from 8.2.2 to 8.2.3.
- [Release notes](https://github.com/DefinitelyTyped/DefinitelyTyped/releases)
- [Commits](https://github.com/DefinitelyTyped/DefinitelyTyped/commits/HEAD/types/mocha)

---
updated-dependencies:
- dependency-name: "@types/mocha"
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Chores: Bump @types/node from 16.0.0 to 16.3.1 (#13)

Bumps [@types/node](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/HEAD/types/node) from 16.0.0 to 16.3.1.
- [Release notes](https://github.com/DefinitelyTyped/DefinitelyTyped/releases)
- [Commits](https://github.com/DefinitelyTyped/DefinitelyTyped/commits/HEAD/types/node)

---
updated-dependencies:
- dependency-name: "@types/node"
  dependency-type: direct:development
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

Oaphi committed:
made addUserInfo more modular; fixed bug showing John Skeet as author of all community wiki posts

Oaphi committed:
Merge branch 'cleanup' of github.com:userscripters/SE-AutoReviewComments into cleanup

Oaphi committed:
moved tabs to own view; curried view switcher

Oaphi committed:
removed margin from actions view

Oaphi committed:
Import/export updates:
- removed separator between JSONP & cancel buttons
- styled 'cancel' button as dangerous

Oaphi committed:
migrated textarea elements to Stacks

Oaphi committed:
removed search input visibility toggling

Oaphi committed:
Info link updates:
- added helper for creating info buttons with Stacks
- switched 'info' button to be a clickable SVG element
- moved info button to the tabs view

Oaphi committed:
remote comments autofetch both JSON and JSON; order of override: JSONP -> JSON -> custom

Oaphi committed:
optimized action lookup by adding a helper

Oaphi committed:
added settings view; moved reset button to settings; fixed bug with postType persisting across events

Oaphi committed:
reworked comment import (legacy from ARC), made it less error-prone

Oaphi committed:
got rid of useless submit button

Oaphi committed:
various optimizations and bug fixes for the welcome view

dependabot[bot] committed:
Chores: Bump @types/node from 16.3.1 to 16.3.3 (#17)

Bumps [@types/node](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/HEAD/types/node) from 16.3.1 to 16.3.3.
- [Release notes](https://github.com/DefinitelyTyped/DefinitelyTyped/releases)
- [Commits](https://github.com/DefinitelyTyped/DefinitelyTyped/commits/HEAD/types/node)

---
updated-dependencies:
- dependency-name: "@types/node"
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Chores: Bump eslint from 7.30.0 to 7.31.0 (#16)

Bumps [eslint](https://github.com/eslint/eslint) from 7.30.0 to 7.31.0.
- [Release notes](https://github.com/eslint/eslint/releases)
- [Changelog](https://github.com/eslint/eslint/blob/master/CHANGELOG.md)
- [Commits](https://github.com/eslint/eslint/compare/v7.30.0...v7.31.0)

---
updated-dependencies:
- dependency-name: eslint
  dependency-type: direct:development
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

Oaphi committed:
added dependency autoupdater

github-actions[bot] committed:
2021-07-19 Automated dependencies update

Oaphi committed:
moved description toggle to settings

Oaphi committed:
Switched to Stacks way of showing/hiding elements

Oaphi committed:
Merge branch 'cleanup' of github.com:userscripters/SE-AutoReviewComments into cleanup

double-beep committed:
Rewrite Cthulhus

double-beep committed:
Merge branch 'cleanup' of https://github.com/userscripters/SE-AutoReviewComments into cleanup

double-beep committed:
npm run build

double-beep committed:
No need to remove dist on build

Oaphi committed:
removed unneeded style overrides

double-beep committed:
FIX: guard against unscheme URL param being null/undefined

Oaphi committed:
reworked icon button factory; made see-through an icon button; removed unused actions view

Oaphi committed:
rebuilt dist

double-beep committed:
Merge branch 'cleanup' of https://github.com/userscripters/SE-AutoReviewComments into cleanup

double-beep committed:
Add close button

double-beep committed:
Do not provide a "JSONP" option

double-beep committed:
Switch to StackExchange.helpers.showToast()

double-beep committed:
FIX: remove d-none on ARC popup if it's been added before

double-beep committed:
FIX: enhance spacing in the import/export tab

Oaphi committed:
bumped dependencies

Oaphi committed:
enabled generation of pretty headers

Oaphi committed:
fixed bug with greeting not being whitespaced

Oaphi committed:
added variable replacer

Oaphi committed:
wrapped script execution into a load event listener

Oaphi committed:
added guard against missing StackExchange global object

Oaphi committed:
bumped dependencies

Oaphi committed:
switched to using Stack Exchange global types package

Oaphi committed:
updated GitHub workflows

Oaphi committed:
improved MD -> HTML regex; switched to tagged template literals

Oaphi committed:
improved preview handling

Oaphi committed:
fixed greeting not being removed when entering edit mode if it contained user variables

Oaphi commented:
All right, we should be in a pretty good shape now except for the following:

- misaligned (relative to the updated to UI that is) "close" icon
- settings view opening first for some reason
- legacy of the original ARC: the need to actually "select" one of the comments before clicking on the insert element (we can use the element to determine which comment user intended to insert instead)
- clean up styling of the "settings" view

Oaphi committed:
made ARC even more robust by exiting early in case StackExchange is not defined at all or if .ready method is not available

Oaphi committed:
fixes #36

Oaphi committed:
bumped dependencies, made package name pretty-printable in userscript headers

Oaphi committed:
fixes #36 completely (does not show last view)

Oaphi committed:
fixes #37

Oaphi committed:
solves welcome input issue of #38

Oaphi committed:
refinements for #37

Oaphi committed:
solves comment search input issue of #38

Oaphi committed:
removes active view switcher bug from #39

Oaphi committed:
simplifies comment insertion procedure #39

Oaphi committed:
minor styling updates

Oaphi committed:
switched to API types package from custom types

Oaphi committed:
slightly tweaked comment list styles

double-beep committed:
Guard against shortened reputation number being x.0k

Oaphi committed:
bumped dependencies

Oaphi committed:
switched to 'all' option when generating headers

Oaphi committed:
switched se-draggable-target CSS class to full popup

Oaphi committed:
linted source

Oaphi committed:
switched draggable handle to the tabs view
as SE's code prevents focus event from being processed

dependabot[bot] committed:
Chores: Bump @types/node from 16.7.10 to 17.0.5 (#78)

Bumps [@types/node](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/HEAD/types/node) from 16.7.10 to 17.0.5.
- [Release notes](https://github.com/DefinitelyTyped/DefinitelyTyped/releases)
- [Commits](https://github.com/DefinitelyTyped/DefinitelyTyped/commits/HEAD/types/node)

---
updated-dependencies:
- dependency-name: "@types/node"
  dependency-type: direct:development
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Chores: Bump @types/mocha from 8.2.3 to 9.0.0 (#18)

Bumps [@types/mocha](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/HEAD/types/mocha) from 8.2.3 to 9.0.0.
- [Release notes](https://github.com/DefinitelyTyped/DefinitelyTyped/releases)
- [Commits](https://github.com/DefinitelyTyped/DefinitelyTyped/commits/HEAD/types/mocha)

---
updated-dependencies:
- dependency-name: "@types/mocha"
  dependency-type: direct:development
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Chores: Bump @types/chai from 4.2.21 to 4.3.0 (#69)

Bumps [@types/chai](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/HEAD/types/chai) from 4.2.21 to 4.3.0.
- [Release notes](https://github.com/DefinitelyTyped/DefinitelyTyped/releases)
- [Commits](https://github.com/DefinitelyTyped/DefinitelyTyped/commits/HEAD/types/chai)

---
updated-dependencies:
- dependency-name: "@types/chai"
  dependency-type: direct:development
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Chores: Bump typescript from 4.5.4 to 4.5.5 (#84)

Bumps [typescript](https://github.com/Microsoft/TypeScript) from 4.5.4 to 4.5.5.
- [Release notes](https://github.com/Microsoft/TypeScript/releases)
- [Commits](https://github.com/Microsoft/TypeScript/compare/v4.5.4...v4.5.5)

---
updated-dependencies:
- dependency-name: typescript
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Chores: Bump eslint from 7.32.0 to 8.7.0 (#83)

Bumps [eslint](https://github.com/eslint/eslint) from 7.32.0 to 8.7.0.
- [Release notes](https://github.com/eslint/eslint/releases)
- [Changelog](https://github.com/eslint/eslint/blob/main/CHANGELOG.md)
- [Commits](https://github.com/eslint/eslint/compare/v7.32.0...v8.7.0)

---
updated-dependencies:
- dependency-name: eslint
  dependency-type: direct:development
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

Oaphi committed:
updated generate-headers package & added meta sites to the list of matches

Oaphi committed:
Merge branch 'cleanup' of github.com:userscripters/SE-AutoReviewComments into cleanup

Oaphi committed:
fixed .action-selected class not being removed from inactive comments after quick insert

Oaphi committed:
made number of textarea rows configurable

Oaphi committed:
bumped generate-headers package to 2.10.1

Oaphi committed:
comment textareas should dynamically resize based on the number of lines in the text

Oaphi committed:
switched to HTMLButtonElement-based implementation of autolinks

Oaphi committed:
fixed edit summary autolink injection

Oaphi committed:
fixed custom close reason autolink injection

Oaphi committed:
bumped generate-headers to 2.11.0 and ensured ARC also matches /review pages

Oaphi committed:
fixed review editor autolink injection

Oaphi committed:
centered the popup to ensure it does not load in the top left corner of the viewport

Oaphi committed:
fixed comment loading in edit mode

Oaphi committed:
improved OP name getter (now works in review queues)

Oaphi committed:
made button styling consistent

Oaphi commented:
@double-beep O.K., I think we are finally ready to go - I straightened out most of the issues, but wanted to get another pair of eyes on the PR before merging. If it looks good to me, I am going to proceed with the merge, update the README, and write up a Stack Apps post about the updated version of the ARC.

double-beep committed:
update dependencies

dependabot[bot] committed:
Chores: Bump mocha from 9.2.2 to 10.0.0 (#117)

Bumps [mocha](https://github.com/mochajs/mocha) from 9.2.2 to 10.0.0.
- [Release notes](https://github.com/mochajs/mocha/releases)
- [Changelog](https://github.com/mochajs/mocha/blob/master/CHANGELOG.md)
- [Commits](https://github.com/mochajs/mocha/compare/v9.2.2...v10.0.0)

---
updated-dependencies:
- dependency-name: mocha
  dependency-type: direct:development
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

Oaphi committed:
fixed headers.sh

Oaphi committed:
only add buttons once per page load (closes #118)

Oaphi committed:
reworked search (closes #122)

Oaphi committed:
major rework:
- input box is determined on the fly (fixes #123);
- fixed flow for determining the type of the post;

Oaphi committed:
reworked new user state management (fixes #124)

Oaphi committed:
updated SE global types package and removed no longer needed @ts-expect-error

Oaphi commented:
Phew, @double-beep , I fixed a heap of issues today. Seems good to go to me but I'll wait on your review before merging.

Oaphi committed:
added <code> element generation for MD -> HTML conversions (closes #126); improved HTML escaping

Oaphi committed:
added setting for toggling ARC debug mode

Oaphi committed:
do not save comments fetched from remote as HTML

Oaphi committed:
minor fix for the getUserInfo helper

Oaphi committed:
reworked injector to use mutation observer

Oaphi committed:
improved matching of MD italic to avoid parsing inline code (#128)

Oaphi committed:
after editing the comments in the import/export view, they should not be saved as HTML

Oaphi committed:
improved matching of bold Markdown to avoid parsing inline code (closes #128)

Oaphi committed:
removed stray console.log

Oaphi committed:
rebuilt dist

Oaphi committed:
removed text color override that made comment preview text too low-contrast in dark mode

Oaphi committed:
improved view switcher to be able to highlight the currently selected tab programmatically

Oaphi committed:
do not save comment as HTML when editing from the filter view

Oaphi committed:
Stacks toggles should be able to have levers on the left; styling adjustments for the settings view

Oaphi committed:
updated quick insert button styling to use Stacks

Oaphi committed:
ensured quick insert buttons are re-shown after closing the edit mode

Oaphi committed:
removed potentially dangerous definite assertions

Oaphi committed:
post edit summary locator should wait until buttons appear in the DOM

Oaphi committed:
closure locator should wait until buttons appear in the DOM

Oaphi committed:
do not d-none class when just making a comment item

Oaphi committed:
added hasMatching method to Store class

Oaphi committed:
save Store class method should return operation status

Oaphi committed:
added interop with the original ARC (closes #130)

Oaphi committed:
ensure JSON mode cannot accidentally try to save comments on edit

Oaphi committed:
added ability to switch back from JSON export

Oaphi commented:
Yet another heap of issues fixed and improvements made (including original ARC interop). Phew, I think *now* it's ready (with one issue outstanding regarding Safari support).

Oaphi committed:
do not attempt to fetch and warn if JSON/JSONP remote is not set

Oaphi committed:
reworked comment targets system

Oaphi committed:
fixed original ARC interop (parsing)

Oaphi committed:
make sure WelcomeMessage from old ARC is not parsed

Oaphi committed:
pre-release cleanup

Oaphi committed:
made updateImpExpComments safer

Oaphi committed:
ensured Store.load always have defaults; do not stringify Store.save if the value is a string already

Oaphi committed:
minor JSDoc cleanup

Oaphi commented:
A heap of issues a day make Oleg go crazy.

---

Joking aside, ironed out a lot regarding storage handling, tightened defaults, and cleaned up the codebase.

double-beep committed:
navigation should be done with s-navigation

Oaphi committed:
do not save default comments as HTML

Oaphi committed:
Merge branch 'cleanup' of github.com:userscripters/SE-AutoReviewComments into cleanup

Oaphi committed:
rebuilt dist

double-beep committed:
switch to s-navigation-tablist controller

double-beep committed:
Merge branch 'cleanup' of https://github.com/userscripters/SE-AutoReviewComments into cleanup

dependabot[bot] committed:
Chores: Bump @types/node from 17.0.30 to 18.0.1 (#148)

Bumps [@types/node](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/HEAD/types/node) from 17.0.30 to 18.0.1.
- [Release notes](https://github.com/DefinitelyTyped/DefinitelyTyped/releases)
- [Commits](https://github.com/DefinitelyTyped/DefinitelyTyped/commits/HEAD/types/node)

---
updated-dependencies:
- dependency-name: "@types/node"
  dependency-type: direct:development
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

Oaphi committed:
added <code> tag to backticks conversion to HTMLtoMarkdown (fixes #223)

double-beep commented:
Almost 2 years! Perhaps it's time to add a bundler and split the code?

Will (hopefully) review this weekend.

double-beep commented:
> (I am happy with any choice)

Great!

I trust your expertise :) - go ahead when you find some time.
